### PR TITLE
Add AppModule context

### PR DESCRIPTION
**Pull request by defunctzombie**
**User-Facing Changes**
None

**Description**
The AppModule context provides a way of replacing or injecting certain app components.

### CODE REVIEW

File: packages/studio-base/src/components/CreateEventDialog.tsx [hunk 1]
```diff
@@ -142,19 +144,33 @@ export function CreateEventDialog(props: { deviceId: string; onClose: () => void
...
+        timestamp: event.startTime.toISOString(),
+        durationNanos: toNanoSec(
+          event.durationUnit === "sec"
+            ? { sec: event.duration, nsec: 0 }
+            : { sec: 0, nsec: event.duration },
+        ).toString(),
+        metadata: keyedMetadata,
+      });
+    } else {
+      await consoleApi.createEvent({
```
Comment (jtbandes) [inline packages/studio-base/src/components/CreateEventDialog.tsx:160]: Why keep `else`? I thought the goal was to remove the consoleApi.

File: packages/studio-base/src/context/AppModuleContext.ts [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+interface IAppModule {
+  providers?: readonly JSX.Element[];
+  syncAdapters?: readonly JSX.Element[];
+  createEvent?: (args: {
+    deviceId: string;
+    timestamp: string;
+    durationNanos: string;
+    metadata: Record<string, string>;
+  }) => Promise<void>;
```
Comment (jhurliman) [inline packages/studio-base/src/context/AppModuleContext.ts:15]: `createEvent` seems like a very specific API. Is the idea to expand the `IAppModule` API surface over time as more functionality is added to Studio?

File: packages/studio-base/src/components/CreateEventDialog.tsx (hunk 1 above)
Comment (defunctzombie) [reply packages/studio-base/src/components/CreateEventDialog.tsx:160]: Indeed that is the goal but I want to have things always in a working state during the transition.

File: packages/studio-base/src/context/AppModuleContext.ts (hunk 2 above)
Comment (jtbandes) [reply packages/studio-base/src/context/AppModuleContext.ts:15]: return of the `GlobalHooks`/`GlobalConfig` https://github.com/foxglove/studio/blob/05f25e2698cb289c3f91696872b6e1a8e9f2e6f3/app/GlobalConfig.ts

Comment (defunctzombie) [reply packages/studio-base/src/context/AppModuleContext.ts:15]: Possibly - at this stage this was the only thing I identified as needing a "shim" to be able to extract ConsoleApi. Maybe IAppModule expands or maybe it stays more opaque via the providers array.

**Comment by defunctzombie** [comment]
I will re-work this PR to see if I can simplify the top level providers out.

**Comment by defunctzombie** [comment]
Replacing this with: https://github.com/foxglove/studio/pull/5361

**Pull request closed without merging**